crates/bundles/src/itar.rs

```diff
@@ -3,198 +3,293 @@
 
 //! The web-friendly "indexed tar" bundle backend.
 //!
-//! The main type offered by this module is the [`IndexedTarBackend`] struct,
-//! which cannot be used directly as a [`tectonic_io_base::IoProvider`] but is
-//! the default backend for cached web-based bundle access through the
-//! [`crate::cache::CachingBundle`] framework.
+//! The main type offered by this module is the [`ItarBundle`] struct,
+//! which can (but should not) be used directly as any other bundle.
+//!
+//! Instead, wrap it in a [`crate::BundleCache`] for filesystem-backed
+//! caching.
 //!
 //! While the on-server file format backing the “indexed tar” backend is indeed
 //! a standard `tar` file, as far as the client is concerned, this backend is
 //! centered on HTTP byte-range requests. For each file contained in the backing
 //! resource, the index file merely contains a byte offset and length that are
 //! then used to construct an HTTP Range request to obtain the file as needed.
 
+use crate::{Bundle, CachableBundle, FileIndex, FileInfo, NET_RETRY_ATTEMPTS, NET_RETRY_SLEEP_MS};
 use flate2::read::GzDecoder;
-use std::{convert::TryInto, io::Read, str::FromStr};
+use std::{
+    collections::HashMap,
+    io::{BufRead, BufReader, Cursor, Read},
+    str::FromStr,
+    thread,
+    time::Duration,
+};
 use tectonic_errors::prelude::*;
 use tectonic_geturl::{DefaultBackend, DefaultRangeReader, GetUrlBackend, RangeReader};
-use tectonic_io_base::digest::{self, DigestData};
-use tectonic_status_base::{tt_note, tt_warning, StatusBackend};
+use tectonic_io_base::{digest, InputHandle, InputOrigin, IoProvider, OpenResult};
+use tectonic_status_base::{tt_note, tt_warning, NoopStatusBackend, StatusBackend};
 
-use crate::cache::{BackendPullData, CacheBackend};
+/// The internal file-information struct used by the [`ItarBundle`].
+#[derive(Clone, Debug)]
+pub struct ItarFileInfo {
+    name: String,
+    offset: u64,
+    length: usize,
+}
 
-const MAX_HTTP_ATTEMPTS: usize = 4;
+impl FileInfo for ItarFileInfo {
+    fn name(&self) -> &str {
+        return &self.name;
+    }
+    fn path(&self) -> &str {
+        return &self.name;
+    }
+}
 
-/// The internal file-information struct used by the [`IndexedTarBackend`].
-#[derive(Clone, Copy, Debug)]
-pub struct FileInfo {
-    offset: u64,
-    length: u64,
+/// A simple FileIndex for compatiblity with [`crate::BundleCache`]
+pub struct ItarFileIndex {
+    content: HashMap<String, ItarFileInfo>,
 }
 
-/// A simple web-based file backend based on HTTP Range requests.
-///
-/// This type implements the [`CacheBackend`] trait and so can be used for
-/// web-based bundle access thorugh the [`crate::cache::CachingBundle`]
-/// framework.
-#[derive(Debug)]
-pub struct IndexedTarBackend {
-    reader: DefaultRangeReader,
+impl ItarFileIndex {
+    fn new() -> Self {
```

Comment:
An argument-less `new()` method is a signpost that you should be implementing the `Default` trait; I think you can auto-derive Default for this type and `ItarFileIndex`.